src/core/chart-generators.js

```diff
@@ -61,146 +59,200 @@ function fixedPrecision (n, ...rest) { return +n.toFixed(PRECISION);};
+      groupBy = graph.axis.y.groupBy.type;
+      yGroup = graph.axis.y.groupBy.value;
+      yUnits = graph.axis.y.units;
+    } else {
+      // a defined y axis exists, but sorting metadata is missing. Error.
+      throw new Error('Error: unable to add data to y axis ' + graph.axis.y.label.text);
     }
-    else if((!y2Units) || timeseries.units == y2Units) {
-      y2Units = timeseries.units;
-      c3Data.axes[timeseriesName] = "y2";
-    }
-    else {
-      throw new Error("Error: too many data axes required for graph");
+  }
+  // if a second y axis already exists, use it.
+  // TODO: This code is included for completeness, but not currently
+  // reachable - new data are never added to an existing graph -
+  // nor well tested.
+  if (nestedAttributeIsDefined(graph, 'axis', 'y2')) {
+    if (nestedAttributeIsDefined(graph, 'axis', 'y2', 'groupBy')
+        && nestedAttributeIsDefined(graph, 'axis', 'y2', 'units')) {
+      // y axis already defined; sort new data according to it.
+      groupBy = graph.axis.y.groupBy.type;
+      y2Group = graph.axis.y2.groupBy.value;
+      y2Units = graph.axis.y2.units;
+    } else {
+      // a defined y axis exists, but sorting metadata is missing. Error.
+      throw new Error('Error: unable to add data to y axis ' + graph.axis.y2.label.text);
     }
   }
-  
-  //whole-graph display options: axis formatting and tooltip behaviour
-  let c3Axis = {};
-  c3Axis.x = monthlyXAxis;
-  c3Axis.y = formatYAxis(yUnits);
-  if(y2Units) { 
-    c3Axis.y2 = formatYAxis(y2Units);
+
+  for (let i = 0; i < graph.data.columns.length; i++) {
+    const seriesName = graph.data.columns[i][0];
+    // skip series consisting of x-axis labels
+    // or that have an axis already.
+    if (_.isUndefined(graph.data.axes[seriesName])
+        && seriesName !== 'x') {
+      const group = seriesMetadata[seriesName][groupBy];
+      const units = seriesMetadata[seriesName].units;
+      if (!yGroup) {
+        // create new primary y axis.
+        yGroup = group;
+        yUnits = units;
+        graph.data.axes[seriesName] = 'y';
+      } else if (group === yGroup) {
+        // assign to existing primary y axis
+        if (units !== yUnits) {
+          throw new Error('Error: mismatched units for graph axis ' + yGroup);
+        }
+        graph.data.axes[seriesName] = 'y';
+      } else if (!y2Group) {
+        // create new secondary y axis
+        y2Group = group;
+        y2Units = units;
+        graph.data.axes[seriesName] = 'y2';
+      } else if (group === y2Group) {
+        // assign to existing secondary y axis
+        if (units !== y2Units) {
+          throw new Error('Error: mismatched units for graph axis' + y2Group);
+        }
+        graph.data.axes[seriesName] = 'y2';
+      } else {
+        // we already have two axes and this data fits with neither.
+        throw new Error('Error: too many data axes required for graph');
+      }
     }
+  }
```

Comment:
There's a lot of repetition here, both between `y` and `y2` and on both branches of an if statement (axes assigment).
Would this shorter code be clearer? It helped me understand what was going on here:

```js
  function checkUnits(theGroup, theUnits, units) {
    if (theGroup && units !== theUnits) {
      throw new Error('Error: mismatched units for graph axis ' + theGroup);
    }
  }

  const seriesNames = graph.data.columns.map(col => col[0]).filter(name => name !== 'x');
  for (const seriesName of seriesNames) {
    const group = seriesMetadata[seriesName][groupBy];
    const units = seriesMetadata[seriesName].units;

    // If both groups are defined and both are different from group, error.
    if (yGroup && group !== yGroup && y2Group && group !== y2Group) {
      throw new Error('Error: too many data axes required for graph');
    }

    checkUnits(yGroup, yUnits, units);
    checkUnits(y2Group, y2Units, units);

    graph.data.axes[seriesName] = yGroup ? 'y' : 'y2';

    // Assign group and units to y or y2, in that order, if they don't already exist.
    if (!yGroup) {
      yGroup = group;
      yUnits = units;
    } else if (!y2Group) {
      y2Group = group;
      y2Units = units;
    }
  }



```

Reply:
Would a better name for `seriesNames` above be `ySeriesNames`? Is that what the condition `name !== 'x'` produces?